app/models/person.rb

```diff
@@ -50,17 +53,101 @@ class Person < ActiveRecord::Base
 
-  has_one :user, foreign_key: :profile_id
+  has_one :user, foreign_key: :profile_id, autosave: true
 
   validates :division_id, presence: true
   validates :first_name, presence: true
 
+  validate :division_role_valid
+  validate :user_valid
+
+  # Transient attributes to fascilitate user managment
+  attr_writer :owning_division_role
+  attr_accessor :password, :password_confirmation
+
+  before_save :update_full_name
+  before_save :update_user
+  after_save :handle_roles
+  after_save :clean_up_passwords
+
+  def update_full_name
+    self.name = "#{first_name} #{last_name}"
+  end
+
+  # Lazy evaluation getter
+  def owning_division_role
+    @owning_division_role = resolve_owning_division_role unless defined? @owning_division_role
+    @owning_division_role
+  end
+
+  def owning_division_role_label
+    label_key = owning_division_role || 'none'
+    I18n.t("people.roles.#{label_key}")
+  end
+
+  def resolve_owning_division_role
+    if user
+      role = user.roles.find_by(resource_type: 'Division', resource_id: division_id)
+      return role.name.to_sym if role
+    end
+    nil
+  end
+
+  def user_required?
+    has_system_access
+  end
+
+  def division_role_valid
+    if has_system_access && (owning_division_role.blank? ||
+      !VALID_DIVISION_ROLES.include?(owning_division_role.to_sym))
+      errors.add(:owning_division_role, I18n.t("people.shared.invalid_division_role"))
+    end
+  end
 
-  #JE todo: this is a placeholder until we implement an automatic update or decide on different handling around the full name
-  def name
-    "#{first_name} #{last_name}"
+  # Delegates validation to the automatically created user instance.
+  def user_valid
+    if user_required?
+      update_user
+      unless user.valid?
+        user.errors.each do |error,message|
+          errors.add(error,message)
+        end
+      end
+    end
   end
 
+  def update_user
+    if user_required?
+      if user
+        # Updates fields on existing user if previously created
+        user.email = self.email
+        if password.present?
+          user.password = self.password
+          user.password_confirmation = self.password_confirmation
+        end
+      else
+        build_user(
+          email: email,
+          password: password,
+          password_confirmation: password_confirmation
+        )
+      end
+    end
+  end
 
+  def handle_roles
+    old_role = resolve_owning_division_role
+    new_role = owning_division_role.present? ? owning_division_role.to_sym : nil
+    # Invalid roles expected to rejected by validation rules. but avoid cryptic error just in case.
+    raise "Unexpected division role: #{new_role}" if new_role && !VALID_DIVISION_ROLES.include?(new_role)
+    if old_role != new_role
+      user.remove_role old_role, division if old_role
+      user.add_role new_role, division if new_role
+    end
+  end
+
+  def clean_up_passwords
```

Comment:
Ok that's fine.

On Thu, May 19, 2016 at 11:24 AM, Joseph Elliott <notifications@github.com>
wrote:

> In app/models/person.rb
> <https://github.com/sassafrastech/madeline_system/pull/84#discussion_r63899202>
> :
>
> >
> > +  def handle_roles
> > +    old_role = resolve_owning_division_role
> > +    new_role = owning_division_role.present? ? owning_division_role.to_sym : nil
> > +    # Invalid roles expected to rejected by validation rules. but avoid cryptic error just in case.
> > +    raise "Unexpected division role: #{new_role}" if new_role && !VALID_DIVISION_ROLES.include?(new_role)
> > +    if old_role != new_role
> > +      user.remove_role old_role, division if old_role
> > +      user.add_role new_role, division if new_role
> > +    end
> > +  end
> > +
> > +  def clean_up_passwords
>
> I saw similar code in the default devise implementation, so I figured it
> was a reasonably paranoid thing to do.
>
> —
> You are receiving this because you commented.
> Reply to this email directly or view it on GitHub
> <https://github.com/sassafrastech/madeline_system/pull/84/files/4cd7c3742da69348808dcc0253537c94ed299a43#r63899202>
>



-- 
Tom Smyth

Worker-Owner, Sassafras Tech Collective
Specializing in innovative, usable tech for social change
sassafras.coop *·* @sassafrastech

Resident, Touchstone Cohousing
touchstonecohousing.org